### PR TITLE
Ecal Unpacker + Downstream Updates to use CUDA Product 

### PR DESCRIPTION
**Pull request by vkhristenko**
#### PR description:

This is a quick implementation of Ecal RAW data decoding.
following note05_021 and source code.

### CODE REVIEW

File: EventFilter/EcalRawToDigi/interface/DeclsForKernels.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#ifndef EventFilter_EcalRawToDigi_interface_DeclsForKernels_h
+#define EventFilter_EcalRawToDigi_interface_DeclsForKernels_h
```
Comment (makortel) [inline EventFilter/EcalRawToDigi/interface/DeclsForKernels.h]: Will this header be needed outside of this package? Or in `src` in addition to `plugins`? If no, please (eventually) move this file to `plugins`.

Comment (vkhristenko) [reply EventFilter/EcalRawToDigi/interface/DeclsForKernels.h]: yeah, this guy gets included into .cu where kernels + host wrapers for kernels are

Comment (vkhristenko) [reply EventFilter/EcalRawToDigi/interface/DeclsForKernels.h]: and .cu are sitting in the lib (/src)

Comment (makortel) [reply EventFilter/EcalRawToDigi/interface/DeclsForKernels.h]: Are the kernels intended to be used by other packages?

Comment (vkhristenko) [reply EventFilter/EcalRawToDigi/interface/DeclsForKernels.h]: nope, only edm plugins from that package use that lib. will move that into plugins/

**Comment by fwyzard** [comment]
@vkhristenko , I don't know if this was sent to you or not.

**Comment by vkhristenko** [comment]
@amassiro

with the last commit, i added a simple unpacker kernel (1 block per fed). gets all the digis, maps electronics ids to det ids. runs w/o problems (seems like). there is a mismatch in number of output digis. at this point i just used Ecal Digi Collections as output targets just to validate, but this is kinda ugly (see here https://github.com/cms-patatrack/cmssw/pull/368/files#diff-765f1f96802f8919a37666fb67c341b5R209)

next, will figure out the diff and then add optimal kernel decomposition.

**Comment by amassiro** [comment]
@grasph  @dsperka

**Comment by vkhristenko** [comment]
@grasp @dsperka, if there is a chance you guys could tell me 1 thing:
- in the Tower blocks, the number of total channels that you get from that is not the number of channels i see in the EBDigiCollection and EEDigiCollection. I saw it for different reasons, one of them i found to be that sometimes det id is 0 for a particular electronics id. again, according to specificaiton i unpack everything just fine. but then i need to remove some of the channels from the final collections. if you guys know what those conditions are that would be awesome... i'm going thru the src now...

**Comment by vkhristenko** [comment]
the last commit fixed ee mismatches, now eb... i've included
- remove channels with det id raw value of 0
- remove channels (tower blocks) that have bx, lv1 mismatch w.r.t. fed header values

**Comment by grasph** [comment]
> @grasp @dsperka, if there is a chance you guys could tell me 1 thing:
> 
>     * in the Tower blocks, the number of total channels that you get from that is not the number of channels i see in the EBDigiCollection and EEDigiCollection. [....]

Sorry I don't know why you the number of dataframes are not consistent. Could it come from data frames that fails some of the different checks and that are dropped ?

Philippe.

**Comment by vkhristenko** [comment]
> > @grasp @dsperka, if there is a chance you guys could tell me 1 thing:
> > ```
> > * in the Tower blocks, the number of total channels that you get from that is not the number of channels i see in the EBDigiCollection and EEDigiCollection. [....]
> > ```
> 
> Sorry I don't know why you the number of dataframes are not consistent. Could it come from data frames that fails some of the different checks and that are dropped ?
> 
> Philippe.

alright, fixed. matches perfectly (for 100 events), adding validation

**Comment by vkhristenko** [comment]
added validation similar to what is done for ecal multifit.

**Comment by vkhristenko** [comment]
[plots.pdf](https://github.com/cms-patatrack/cmssw/files/3374530/plots.pdf)

**Comment by vkhristenko** [comment]
After testing on a larger sample, i see there are still issues but this is getting close
```
>>> nentries = 6247
*** mismatch in ndigis: ie = 249  ngpuebs = 3765  ncpuebs = 3766  ngpuees = 946  ncpuees = 946
*** mismatch in ndigis: ie = 1572  ngpuebs = 2762  ncpuebs = 2763  ngpuees = 635  ncpuees = 635
*** mismatch in ndigis: ie = 1996  ngpuebs = 2343  ncpuebs = 2344  ngpuees = 811  ncpuees = 811
*** mismatch in ndigis: ie = 2768  ngpuebs = 2959  ncpuebs = 2960  ngpuees = 608  ncpuees = 608
*** mismatch in ndigis: ie = 4873  ngpuebs = 2248  ncpuebs = 2249  ngpuees = 1242  ncpuees = 1242
```

**Comment by vkhristenko** [comment]
- added cuda product usage of produced digis
- added cpu consumer to enqueue transfers and store digis back to edm event for persistance
- validation 100% (above commit earlier fixed the last mismsatch) on ~3K events... good enough for now.
- now will add this for downstream mulfitit producer

**Comment by vkhristenko** [comment]
@fwyzard @felicepantaleo @makortel 

although i did not incorporate what Matti asked for (moving things from src to plugins),
this is ready to be looked at. 

- raw decoding. simple impl for now. dirty (did not remove commented out stuff). but seems to work/validated on a large sample. runs on average at 400us, from nvprof... full results to be measured. can be probably improved substantially
- updated all the products to be done with cuda products + added consumers to produce cpu products
- validation runs fine (for digis and for rec hits). basically chain of raw decoding + multifit runs fine/validated

**Comment by amassiro** [comment]
It would be nice to have, as currently already done for timing calculation, "amplitude error" calculation on gpu, something that we can enable and disable with a flag.
Currently, on cpu, amplitude error is calculated but then not used downstream, but it can be a nice handle to fight noise for run III, so better have this possibility.

**Comment by amassiro** [comment]
@rchatter (ecal reco expert) added in the thread

**Comment by fwyzard** [comment]
Superseded by #443.

**Pull request closed without merging**